lib/bolt/cli.rb

```diff
@@ -379,710 +341,584 @@ def validate(options)
+              executor.subscribe(outputter, %i[message verbose])
+            end
           else
-            list_tasks
+            executor.subscribe(outputter)
           end
-        when 'plan'
-          if options[:object]
-            show_plan(options[:object])
-          else
-            list_plans
+
+          executor.subscribe(log_outputter)
+
+          # TODO: Figure out where this should really go. It doesn't seem to
+          #       make sense in the application, since the params should already
+          #       be data when they reach that point.
+          if %w[plan task].include?(command) && action == 'run'
+            options[:params] = parse_params(
+              command,
+              options[:object],
+              pal,
+              **options.slice(:params, :params_parsed)
+            )
           end
-        when 'inventory'
-          if options[:detail]
-            show_targets
-          else
-            list_targets
+
+          if command == 'script'
+            options[:arguments] = options[:leftovers]
           end
-        when 'group'
-          list_groups
-        when 'module'
-          list_modules
-        when 'plugin'
-          list_plugins
+
+          application = Bolt::Application.new(
+            analytics: analytics,
+            config:    config,
+            executor:  executor,
+            inventory: inventory,
+            pal:       pal,
+            plugins:   plugins
+          )
+
+          process_command(application, command, action, options)
+        ensure
+          analytics&.finish
         end
-        return 0
-      when 'convert'
-        pal.convert_plan(options[:object])
-        return 0
       end
+    end
 
-      message = 'There might be processes left executing on some nodes.'
+    # Process the command.
+    #
+    # @param app [Bolt::Application] The application.
+    # @param command [String] The command.
+    # @param action [String, NilClass] The action.
+    # @param options [Hash] The CLI options.
+    #
+    private def process_command(app, command, action, options)
+      case command
+      when 'apply'
+        results = outputter.spin do
+          app.apply(options[:object], options[:targets], **options.slice(:code, :noop))
+        end
+        rerun.update(results)
+        app.shutdown
+        outputter.print_apply_result(results)
+        results.ok? ? SUCCESS : FAILURE
 
-      if %w[task plan].include?(options[:subcommand]) && options[:task_options] && !options[:params_parsed] && pal
-        options[:task_options] = pal.parse_params(options[:subcommand], options[:object], options[:task_options])
-      end
+      when 'command'
+        outputter.print_head
+        results = outputter.spin do
+          app.command_run(options[:object], options[:targets], **options.slice(:env_vars))
+        end
+        rerun.update(results)
+        app.shutdown
+        outputter.print_summary(results, results.elapsed_time)
```

Comment:
It seems a little funny now that we publish messages to the outputter from the executor, but then for results we *return* the results and pass them to the outputter. I think this is too much work to be in scope for this, but it'd be nice if `with_node_logging` published the results instead of returning them, then the outputter could just see the result and print it according to the type. That'd also be nice because then `app.shutdown` could just be in an `ensure` in this function instead of in every case statement.

Reply:
Hm, I like that idea. We can ticket and discuss another time.